File: backend/app/schemas/divergencia.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, model_validator


StatusDivergencia = Literal["aberta", "em_analise", "resolvida", "ignorada"]

STATUS_VALIDOS: set[str] = {"aberta", "em_analise", "resolvida", "ignorada"}


class DivergenciaAtualizacao(BaseModel):
    status: StatusDivergencia | None = None
    observacao: str | None = None

    @model_validator(mode="after")
    def ao_menos_um_campo(self) -> "DivergenciaAtualizacao":
        if self.status is None and self.observacao is None:
            raise ValueError("Pelo menos um campo deve ser informado: status ou observacao.")
        return self


class DivergenciaDetalhe(BaseModel):
    id: uuid.UUID
    fechamento_id: uuid.UUID
    item_conciliacao_id: uuid.UUID
    tipo_divergencia: str
    severidade: str
    descricao: str
    valor_previsto: Decimal | None
    valor_realizado: Decimal | None
    diferenca_valor: Decimal | None
    data_prevista: date | None
    data_realizada: date | None
    diferenca_dias: int | None
    status: str
    observacao: str | None
    resolvido_em: datetime | None
    atualizado_por_usuario_id: uuid.UUID | None
    criado_em: datetime
    atualizado_em: datetime

    model_config = {"from_attributes": True}

File: frontend/app/types/divergencia.ts
export type StatusDivergencia = 'aberta' | 'em_analise' | 'resolvida' | 'ignorada'

export type TipoDivergencia =
  | 'divergencia_data'
  | 'divergencia_valor'
  | 'previsto_nao_realizado'
  | 'realizado_nao_previsto'
  | 'duplicidade_extrato'
  | 'duplicidade_fluxo'
  | 'pendente_analise_manual'

export type SeveridadeDivergencia = 'baixa' | 'media' | 'alta'

export interface Divergencia {
  id: string
  item_conciliacao_id: string
  tipo_divergencia: TipoDivergencia | string
  severidade: SeveridadeDivergencia | string
  descricao: string
  valor_previsto: number | null
  valor_realizado: number | null
  diferenca_valor: number | null
  data_prevista: string | null
  data_realizada: string | null
  diferenca_dias: number | null
  status: StatusDivergencia
  observacao: string | null
  resolvido_em: string | null
  atualizado_por_usuario_id: string | null
  criado_em: string
  atualizado_em: string
}

export interface AtualizarDivergenciaPayload {
  status?: StatusDivergencia
  observacao?: string
}

export interface FiltroDivergencias {
  status: StatusDivergencia | ''
  tipo_divergencia: string
  severidade: string
  busca: string
}

export const LABELS_STATUS_DIVERGENCIA: Record<string, string> = {
  aberta:     'Aberta',
  em_analise: 'Em análise',
  resolvida:  'Resolvida',
  ignorada:   'Ignorada',
}

export const LABELS_TIPO_DIVERGENCIA: Record<string, string> = {
  divergencia_data:         'Divergência de data',
  divergencia_valor:        'Divergência de valor',
  previsto_nao_realizado:   'Previsto não realizado',
  realizado_nao_previsto:   'Realizado não previsto',
  duplicidade_extrato:      'Duplicidade no extrato',
  duplicidade_fluxo:        'Duplicidade no fluxo',
  pendente_analise_manual:  'Pendente para análise',
}

export const LABELS_SEVERIDADE: Record<string, string> = {
  baixa: 'Baixa',
  media: 'Média',
  alta:  'Alta',
}

export function labelStatusDivergencia(s: string) {
  return LABELS_STATUS_DIVERGENCIA[s] ?? s
}

export function labelTipoDivergencia(t: string) {
  return LABELS_TIPO_DIVERGENCIA[t] ?? t
}

export function labelSeveridade(s: string) {
  return LABELS_SEVERIDADE[s] ?? s
}
